src/core/execute.c

```diff
@@ -3823,6 +3831,53 @@ static bool insist_on_sandboxing(
         return false;
 }
 
+static int setup_ephemeral(const ExecContext *context, ExecRuntime *runtime) {
+        _cleanup_close_ int fd = -EBADF;
+        int r;
+
+        if (!runtime || !runtime->ephemeral)
+                return 0;
+
+        r = posix_lock(runtime->ephemeral_storage_socket[0], LOCK_EX);
+        if (r < 0)
+                return log_debug_errno(r, "Failed to lock ephemeral storage socket: %m");
+
+        CLEANUP_POSIX_UNLOCK(runtime->ephemeral_storage_socket[0]);
+
+        fd = receive_one_fd(runtime->ephemeral_storage_socket[0], MSG_PEEK|MSG_DONTWAIT);
+        if (fd >= 0)
+                /* We got an fd! That means ephemeral has already been set up, so nothing to do here. */
+                return 0;
+
+        if (fd != -EAGAIN)
+                return log_debug_errno(fd, "Failed to receive file descriptor queued on ephemeral storage socket: %m");
+
+        log_debug("Making ephemeral snapshot of %s to %s",
+                  context->root_image ?: context->root_directory, runtime->ephemeral);
+
+        if (context->root_image)
+                fd = copy_file_full(context->root_image, runtime->ephemeral, O_EXCL, 0600,
+                                    FS_NOCOW_FL, FS_NOCOW_FL,
```

Comment:
uh? that forces an immediate sector copy no? how does this fit together with COPY_REFLINK?

Reply:
I copied this from nspawn if I remember correctly, but I also just found out in mkosi that this doesn't actually work, we get EINVAL from FICLONE if we disable COW before doing the reflink. So I'll rework this to not disable COW. Not going to be ideal for performance though.

Reply:
i think kernel changed behaviour there, hnce the nspawn behaviour.

maybe we should turn on FS_NOCOW_FL *after* doing the copy. This won't impress btrfs much, but communicates clearly what kind of operation we actually want once the initial reflink is done.